include/openssl/kdf.h

```diff
@@ -15,80 +15,104 @@
 extern "C" {
 #endif
 
-# define EVP_PKEY_CTRL_TLS_MD                   (EVP_PKEY_ALG_CTRL)
+/* Generic controls */
+# define EVP_PKEY_CTRL_KDF_MD                   (EVP_PKEY_ALG_CTRL)
+# define EVP_PKEY_CTRL_KDF_PASS                 (EVP_PKEY_ALG_CTRL + 8)
+# define EVP_PKEY_CTRL_KDF_SALT                 (EVP_PKEY_ALG_CTRL + 4)
+# define EVP_PKEY_CTRL_KDF_KEY                  (EVP_PKEY_ALG_CTRL + 5)
+# define EVP_PKEY_CTRL_KDF_INFO                 (EVP_PKEY_ALG_CTRL + 6)
+# define EVP_PKEY_CTRL_KDF_MODE                 (EVP_PKEY_ALG_CTRL + 7)
+
+/* KDF-specific controls */
 # define EVP_PKEY_CTRL_TLS_SECRET               (EVP_PKEY_ALG_CTRL + 1)
 # define EVP_PKEY_CTRL_TLS_SEED                 (EVP_PKEY_ALG_CTRL + 2)
-# define EVP_PKEY_CTRL_HKDF_MD                  (EVP_PKEY_ALG_CTRL + 3)
-# define EVP_PKEY_CTRL_HKDF_SALT                (EVP_PKEY_ALG_CTRL + 4)
-# define EVP_PKEY_CTRL_HKDF_KEY                 (EVP_PKEY_ALG_CTRL + 5)
-# define EVP_PKEY_CTRL_HKDF_INFO                (EVP_PKEY_ALG_CTRL + 6)
-# define EVP_PKEY_CTRL_HKDF_MODE                (EVP_PKEY_ALG_CTRL + 7)
-# define EVP_PKEY_CTRL_PASS                     (EVP_PKEY_ALG_CTRL + 8)
-# define EVP_PKEY_CTRL_SCRYPT_SALT              (EVP_PKEY_ALG_CTRL + 9)
 # define EVP_PKEY_CTRL_SCRYPT_N                 (EVP_PKEY_ALG_CTRL + 10)
 # define EVP_PKEY_CTRL_SCRYPT_R                 (EVP_PKEY_ALG_CTRL + 11)
 # define EVP_PKEY_CTRL_SCRYPT_P                 (EVP_PKEY_ALG_CTRL + 12)
 # define EVP_PKEY_CTRL_SCRYPT_MAXMEM_BYTES      (EVP_PKEY_ALG_CTRL + 13)
 
+/* Aliases for backwards-compatibility */
+# define EVP_PKEY_CTRL_TLS_MD                   EVP_PKEY_CTRL_KDF_MD
+# define EVP_PKEY_CTRL_HKDF_MD                  EVP_PKEY_CTRL_KDF_MD
```

Comment:
You need to make sure that `pkey_hkdf_ctrl` in `crypto/kdf/hkdf.c` *also* understands `(EVP_PKEY_ALG_CTRL + 3)`.  

You could do that by defining `EVP_PKEY_CTRL_HKDF_MD` to be that value, and in that case, you'll need this patch:

``` diff
diff --git a/crypto/kdf/hkdf.c b/crypto/kdf/hkdf.c
index 63c3523c27..058d498879 100644
--- a/crypto/kdf/hkdf.c
+++ b/crypto/kdf/hkdf.c
@@ -72,6 +72,7 @@ static int pkey_hkdf_ctrl(EVP_PKEY_CTX *ctx, int type, int p1, void *p2)
 
     switch (type) {
     case EVP_PKEY_CTRL_HKDF_MD:
+    case EVP_PKEY_CTRL_KDF_MD:
         if (p2 == NULL)
             return 0;
 
```

**OR** you could define another macro, like this:

``` C
# define EVP_PKEY_CTRL_OLD_HKDF_MD              (EVP_PKEY_ALG_CTRL + 3)
```

and the following diff:

``` diff
diff --git a/crypto/kdf/hkdf.c b/crypto/kdf/hkdf.c
index 63c3523c27..b6d4bfef57 100644
--- a/crypto/kdf/hkdf.c
+++ b/crypto/kdf/hkdf.c
@@ -71,7 +71,8 @@ static int pkey_hkdf_ctrl(EVP_PKEY_CTX *ctx, int type, int p1, void *p2)
     HKDF_PKEY_CTX *kctx = ctx->data;
 
     switch (type) {
-    case EVP_PKEY_CTRL_HKDF_MD:
+    case EVP_PKEY_CTRL_OLD_HKDF_MD:
+    case EVP_PKEY_CTRL_KDF_MD:
         if (p2 == NULL)
             return 0;
 
```

Either way, *something* like that is needed, or we have a backward compatibility problem.